pkg/scheduler/plugins/gpushare/plugin.go

```diff
@@ -0,0 +1,285 @@
+	// nodeLister      corelisters.NodeLister
+	nodeDeviceCache *nodeDeviceCache
+}
+
+var (
+	_ framework.PreFilterPlugin = &Plugin{}
+	_ framework.FilterPlugin    = &Plugin{}
+	_ framework.ReservePlugin   = &Plugin{}
+	_ framework.PreBindPlugin   = &Plugin{}
+)
+
+type preFilterState struct {
+	skip              bool
+	allocResult       []*apiext.DeviceAllocation
+	convertedResource corev1.ResourceList
+}
+
+func (s *preFilterState) Clone() framework.StateData {
+	return s
+}
+
+func (g *Plugin) Name() string {
+	return Name
+}
+
+func (g *Plugin) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	state := &preFilterState{skip: true}
+
+	podRequest, _ := resource.PodRequestsAndLimits(pod)
+
+	if isGPUPod(podRequest) {
+		combination, err := validateGPURequest(podRequest)
+		if err != nil {
+			return framework.NewStatus(framework.Error, err.Error())
+		}
+		state.skip = false
+		state.convertedResource = convertGPUResource(podRequest, combination)
+	}
+	cycleState.Write(stateKey, state)
+	return nil
+}
+
+func (g *Plugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
+	c, err := cycleState.Read(stateKey)
+	if err != nil {
+		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
+		return nil, fmt.Errorf("reading %q from cycleState: %w", stateKey, err)
+	}
+
+	s, ok := c.(*preFilterState)
+	if !ok {
+		return nil, fmt.Errorf("%+v  convert to gpuShare.preFilterState error", c)
+	}
+	return s, nil
+}
+
+func (g *Plugin) Filter(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	state, err := getPreFilterState(cycleState)
+	if err != nil {
+		return framework.AsStatus(err)
+	}
+	if state.skip {
+		return nil
+	}
+
+	if nodeInfo.Node() == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	nodeDeviceInfo := g.nodeDeviceCache.getNodeDevice(nodeInfo.Node().Name)
+	if nodeDeviceInfo == nil {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrMissingDevice)
+	}
+
+	nodeDeviceInfo.lock.Lock()
+	defer nodeDeviceInfo.lock.Unlock()
+	nodeDeviceTotal := nodeDeviceInfo.deviceTotal[schedulingv1alpha1.GPU]
+	// node does not have GPU
+	if len(*nodeDeviceTotal) <= 0 {
+		return framework.NewStatus(framework.Unschedulable, ErrNoMatchNode)
+	}
+
+	podRequest := state.convertedResource
+
+	if gpuMem, ok := podRequest[apiext.GPUMemory]; ok {
+		podRequest[apiext.GPUMemoryRatio] = memBytesToRatio(gpuMem, (*nodeDeviceTotal)[0][apiext.GPUMemory])
+	} else {
+		gpuMemRatio := podRequest[apiext.GPUMemoryRatio]
+		podRequest[apiext.GPUMemory] = memRatioToBytes(gpuMemRatio, (*nodeDeviceTotal)[0][apiext.GPUMemory])
+	}
+
+	nodeFreeGPUResource := nodeDeviceInfo.deviceFree[schedulingv1alpha1.GPU]
+
+	if checkNodeGPUSatisfied(podRequest, nodeFreeGPUResource) {
```

Comment:
checkNodeGPUSatisfied should be nodeDevice logic, because nodeFreeGPUResource is direct-reference of nodeDeviceInfo, if we modify nodeFreeGPUResource expectlly\unexpectlly,  it will be wired to modify a class varible out of class's function(may confict with object-based-programming habit). anyway, expose the direct-reference of  nodeDeviceInfo is not recommended.